src/utils/csv.ts

```diff
@@ -0,0 +1,139 @@
+import { saveAs } from 'file-saver';
+import moment from 'moment-timezone';
+import { getGranularityInMS } from './utils';
+
+export interface DatapointsToCSVProps {
+  data: DatapointsGetAggregateDatapoint[];
+  granularity: string;
+  aggregate?: Aggregate;
+  delimiter?: Delimiters;
+  format?: string;
+}
+
+export enum Delimiters {
+  Coma = ',',
+  Tab = '\t',
+  Semicolon = ';',
+}
+
+export function datapointsToCSV({
+  data,
+  granularity,
+  aggregate = 'average',
+  delimiter = Delimiters.Coma,
+  format,
+}: DatapointsToCSVProps): string {
+  const ids = data.map(({ id, externalId }) => externalId || id.toString());
+
+  const arrangedData = arrangeDatapointsByTimestamp({
+    data,
+    aggregate,
+    granularity,
+  });
+
+  const csvStrings = generateCSVStringsArray(arrangedData, delimiter, format);
+
+  return [['timestamp', ...ids].join(delimiter), ...csvStrings].join('\n');
+}
+
+export function downloadCSV(src: string, filename: string = 'data.csv') {
+  const blob = new Blob([src], { type: 'text/csv;charset=utf-8;' });
+
+  saveAs(blob, filename);
+}
+
+function getStartTimestampAndIterationsTotal(
+  data: DatapointsGetAggregateDatapoint[]
+): [number, number] {
+  let startTimestamp = Infinity;
+  const iterationsTotal = data.reduce((acc, { datapoints }) => {
+    if (datapoints.length) {
+      const timestamp = datapoints[0].timestamp.getTime();
+
+      if (startTimestamp > timestamp) {
+        startTimestamp = timestamp;
+      }
+    }
+
+    return datapoints.length + acc;
+  }, 0);
+
+  return [startTimestamp, iterationsTotal];
+}
+
+function arrangeDatapointsByTimestamp({
+  data,
+  aggregate,
+  granularity: granularityString,
+}: Omit<DatapointsToCSVProps, 'delimiter'>): (number | string)[][] {
+  if (!granularityString) {
+    return [];
+  }
+
+  const arrangedData = [];
+  const granularity = getGranularityInMS(granularityString);
+
+  const pointers = Array(data.length).fill(0);
+  const [startTimestamp, iterationsTotal] = getStartTimestampAndIterationsTotal(
+    data
+  );
+
+  let iterationLeft = iterationsTotal;
+  let currentTimestamp = startTimestamp;
+
+  for (let iteration = 0; iteration <= iterationsTotal; iteration++) {
+    let valuesGet = 0;
+    const values: string[] = [];
+
+    data.forEach(({ datapoints }, index) => {
+      const dpPointer = pointers[index];
+      const datapoint = datapoints[dpPointer];
+
+      if (datapoint) {
+        const timestamp = datapoints[dpPointer].timestamp.getTime();
+
+        if (timestamp === currentTimestamp) {
+          const value = datapoints[dpPointer][aggregate!];
+
+          values.push(value!.toString());
+          valuesGet++;
+          pointers[index]++;
+          iterationLeft--;
+
+          return;
+        }
+      }
+
+      values.push('');
+    });
+
+    iteration++;
+
+    if (valuesGet) {
+      arrangedData.push([currentTimestamp, ...values]);
+    }
+
+    currentTimestamp += granularity;
+    iterationLeft -= valuesGet;
```

Comment:
how does this work? 
you do 
```
valuesGet++;
iterationLeft--;
```
in for loop, so you already subtracted values get. why do you subtract it again?
I guess that's because you don't use iterations left properly here
you should add it as a for-loop break check on top `for(let iteration = 0; iteration <= iterationLeft; iteration++)`, not to have explicit `if (iterationLeft <= 0)` below

Reply:
if you do that then you don't need to have `valuesGet ` as a number. `hasValues: boolean` is enough

Reply:
having `iterationLeft` and  `iterationsTotal` is a safe way to iterate through datapoints (dps). In case of low granularity of dps (5s for example), some timestamps might be missed. To align all dps we need to run through all of them. But some of them related to the same timestamp, and some is don't. In the worst case dps don't align at all, and you'll need to make `iterationsTotal` number of iterations to align them and in this case `valuesGet` will be always 1 for each step. 
In case of all timestamp presents, your `valuesGet` will always equal to number of timeseries. In this case your total iterations number will be `iterationsTotal / timeseries.length`.
This approach just keep 2 pointers to keep iterating through dps – `iterationsTotal` and `iterationLeft `